### PR TITLE
Support Webpack 4.x

### PR DESCRIPTION
**Pull request by walkerburgin**
Looks like PR #46 is pretty stale. I took a quick crack at updating to support webpack 4.x by going through the schema diff outlined in issue #45. 

I'm currently trying to use `webpack-chain` in a project, but not having support for webpack 4 is a pretty big blocker.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,6 +25,6 @@
   "devDependencies": {
     "ava": "^0.22.0",
     "changelog": "^1.4.0",
-    "webpack": "^3.6.0"
+    "webpack": "^4.5.0"
```
Comment (edmorley) [inline package.json]: Could you update `README.md` to say now supports webpack 2-4 instead of 2-3?

File: test/Config.js [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import test from 'ava';
 import Config from '../src/Config';
-import { validate } from 'webpack';
+import { validateSchema } from 'webpack';
```
Comment (edmorley) [inline test/Config.js]: I know #46 made this change, however I don't think it's correct. From what I can tell at first glance (and there aren't really any API docs), `validateSchema` expects both schema and options, so if called directly without a schema, it's not validating anything.

File: src/Optimization.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+const ChainedMap = require('./ChainedMap');
+
+module.exports = class extends ChainedMap {
```
Comment (edmorley) [inline src/Optimization.js]: Could you add a new test `test/Optimization.js` for this?

File: README.md [hunk 4]
```diff
@@ -347,6 +347,7 @@ config
   .context(context)
   .externals(externals)
   .loader(loader)
+  .mode(mode)
```
Comment (edmorley) [inline README.md]: The config merging section of the README also references the top level options, could you add `mode` there too? (I see `parallelism` is also missing from that, perhaps worth fixing whilst we're there)

Also, could you add the `optimization` options to the README too? :-)

**Comment by edmorley** [comment]
Could you also rebase on master and resolve the conflict? I'd just delete `yarn.lock` and recreate to pick up the latest changes.

File: test/Config.js (hunk 2 above)
Comment (walkerburgin) [reply test/Config.js]: Looks like you're right!

File: src/Output.js [hunk 5]
```diff
@@ -7,11 +7,14 @@ module.exports = class extends ChainedMap {
       'auxiliaryComment',
       'chunkFilename',
       'chunkLoadTimeout',
+      'chunkCallbackName',
```
Comment (eliperelman) [inline src/Output.js]: This is super nitpicky, but could you move this to be before `chunkFilename` so it stays in alphabetical order?

File: test/Config.js [hunk 6]
```diff
@@ -85,14 +85,24 @@ test('toConfig with values', t => {
     .output
       .path('build')
       .end()
+    .mode("development")
```
Comment (eliperelman) [inline test/Config.js]: Can you replace all the double quotes in here with single quotes to match the style? I need to go in and add better linting tools some day. 😄

File: src/Output.js (hunk 5 above)
Comment (edmorley) [reply src/Output.js]: There's a bit of a mix throughout webpack-chain at the moment of alphabetical order vs schema order - would be good to sort it all out at some point :-)

Comment (eliperelman) [reply src/Output.js]: Oh, I thought I already cleaned all those up... 🤔

**Comment by edmorley** [comment]
Travis is failing on the node 6 job with:
`error webpack@4.5.0: The engine "node" is incompatible with this module. Expected version ">=6.11.5".`

Turns out webpack 4 has bumped the minimum node version:
https://github.com/webpack/webpack/blob/v4.5.0/package.json#L72

...could you increase the version specified in `.travis.yml` to match? (I guess we could always just use `6` there instead to always use the latest - but I presume the intention was to check backwards compatibility?)

File: src/Output.js (hunk 5 above)
Comment (walkerburgin) [reply src/Output.js]: Ok, I did a very quick run through to get all of the `extend()` calls at least to specify in alphabetical order.

+1 to setting up linting and/or Prettier 😄

**Comment by eliperelman** [comment]
Our pleasure! Released in v4.6.0.

**Pull request merged**